### ISSUE TITLE
Gray region in output

### ISSUE THREAD

**Issue by Ad-Hua** [jingyuanli001/RFR-Inpainting#21]
Thanks for your great work! I have trained your model with my own dataset. However, I found sometimes there will be a gray region in the output, which is the center region of masked area. The attached images is one example.
![3C74BD47-9E98-2B55-1E63-67FCD6A95C8D20200728_L](https://user-images.githubusercontent.com/73102397/96421154-ff879700-1228-11eb-847e-833ee364bb1e.jpg)
![3C74BD47-9E98-2B55-1E63-67FCD6A95C8D20200728_L](https://user-images.githubusercontent.com/73102397/96421245-1e862900-1229-11eb-919c-3f2b30d488d2.jpg)
![3C74BD47-9E98-2B55-1E63-67FCD6A95C8D20200728_L](https://user-images.githubusercontent.com/73102397/96421393-4d9c9a80-1229-11eb-9e54-9486ed8f2a59.png)

How can I improve this?

**Comment by jingyuanli001** [comment]
Hi,
I believe this is because the masked region given is too large and the model failed to inpaint the whole area before the iteration number runs out, which then leads to the gray region. This is an architectural limitation of the RFR-Net. The best way to solve this issue is to add one or two extra encoding/decoding layers before and after the RFR module respectively while this would require you to retrain the model. If you don't want to change the model, I would suggest you put the unfinished images and masks into the model again, and let the model inpaint on the incomplete results.

**Comment by Ad-Hua** [comment]
Thanks for your help. I will follow your advice to see the results.

**Comment by Ad-Hua** [comment]
Hi! I followed your suggestion but still get the similar result. I add two layers in encoder and two layers in decoder like this:
 (enc_4): Sequential(
    (0): Conv2d(512, 512, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
    (1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (2): ReLU(inplace=True)
  )
  (enc_5): Sequential(
    (0): Conv2d(512, 512, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
    (1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (2): ReLU(inplace=True)
(dec_5): Sequential(
    (0): ConvTranspose2d(1024, 512, kernel_size=(4, 4), stride=(2, 2), padding=(1, 1), bias=False)
    (1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (2): LeakyReLU(negative_slope=0.2, inplace=True)
  )
  (dec_4): Sequential(
    (0): ConvTranspose2d(1024, 512, kernel_size=(4, 4), stride=(2, 2), padding=(1, 1), bias=False)
    (1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (2): LeakyReLU(negative_slope=0.2, inplace=True)
  )

Will I put more RFRModule help?

**Comment by jingyuanli001** [comment]
> Hi! I followed your suggestion but still get the similar result. I add two layers in encoder and two layers in decoder like this:
> (enc_4): Sequential(
> (0): Conv2d(512, 512, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
> (1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
> (2): ReLU(inplace=True)
> )
> (enc_5): Sequential(
> (0): Conv2d(512, 512, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
> (1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
> (2): ReLU(inplace=True)
> (dec_5): Sequential(
> (0): ConvTranspose2d(1024, 512, kernel_size=(4, 4), stride=(2, 2), padding=(1, 1), bias=False)
> (1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
> (2): LeakyReLU(negative_slope=0.2, inplace=True)
> )
> (dec_4): Sequential(
> (0): ConvTranspose2d(1024, 512, kernel_size=(4, 4), stride=(2, 2), padding=(1, 1), bias=False)
> (1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
> (2): LeakyReLU(negative_slope=0.2, inplace=True)
> )
> 
> Will I put more RFRModule help?

Yes, putting more RFR modules definitely helps. The encoding layer in the code above should be composed of Partial convolution layers which updates the mask m, instead of the normal convolution layers which do not update the mask.

**Comment by Ad-Hua** [comment]
Thanks for your help! I tried more Pconv and RFR modules and the problem was solved.

**Comment by qwerdbeta** [comment]
@Ad-Hua can you please share your code? I am trying to add these layers but getting some problems. Thanks!

**Comment by Ad-Hua** [comment]
Hi!
I change the number of RFR modules from 6 to 10 and solve the problem. 
As for adding Pconv, I added like this:

self.Pconv21 = PartialConv2d(64, 64, 7, 1, 3, multi_channel = True, bias = False)
self.Pconv22 = PartialConv2d(64, 64, 7, 1, 3, multi_channel = True, bias = False)
_____________________________________________________________________________________
self.Pconv23 = PartialConv2d(64, 64, 3, 1, 1, multi_channel = True, bias = False)
self.Pconv24 = PartialConv2d(64, 64, 3, 1, 1, multi_channel = True, bias = False)
_____________________________________________________________________________________
This can help to release that problem but cannot solve totally . So I suggest you to add RFR modules if you dont require a high speed inference.

**Comment by qwerdbeta** [comment]
thanks a lot!

**Comment by qwerdbeta** [comment]
@Ad-Hua it works. When I made the change you suggested, it seems to make the filled in area more fine grain. It's still early in the iterations so I'm still waiting to see how it comes out as it converges more. Are our images larger than 256x256? they look large ?

**Comment by Ad-Hua** [comment]
Hi! @qwerdbeta 
I assume you are asking whether the gray region become larger if the input mask is larger. In my opinion, Yes. So you may change some parameters to get good results.

**Comment by qwerdbeta** [comment]
@Ad-Hua No, I meant your target_size as defined in run.py.

Also, I noticed that you can get rid of the grey areas by incrasing the values:
```

        self.Pconv1 = PartialConv2d(3, 64, 7, 2, 3, multi_channel = True, bias = False)
        self.bn1 = nn.BatchNorm2d(64)
        self.Pconv2 = PartialConv2d(64, 64, 7, 1, 3, multi_channel = True, bias = False)
        self.bn20 = nn.BatchNorm2d(64)
        self.Pconv21 = PartialConv2d(64, 64, 7, 1, 3, multi_channel = True, bias = False)
        self.Pconv22 = PartialConv2d(64, 64, 7, 1, 3, multi_channel = True, bias = False)
        self.Pconv23 = PartialConv2d(64, 64, 3, 1, 1, multi_channel = True, bias = False)
        self.Pconv24 = PartialConv2d(64, 64, 3, 1, 1, multi_channel = True, bias = False)
```

to:

        self.Pconv1 = PartialConv2d(3, 64, 7, 2, 3, multi_channel = True, bias = False)
        self.bn1 = nn.BatchNorm2d(64)
        self.Pconv2 = PartialConv2d(64, 64, 9, 1, 4, multi_channel = True, bias = False)
        self.bn20 = nn.BatchNorm2d(64)
        self.Pconv21 = PartialConv2d(64, 64, 9, 1, 4, multi_channel = True, bias = False)
        self.Pconv22 = PartialConv2d(64, 64, 9, 1, 4, multi_channel = True, bias = False)
        self.Pconv23 = PartialConv2d(64, 64, 3, 1, 1, multi_channel = True, bias = False)
        self.Pconv24 = PartialConv2d(64, 64, 3, 1, 1, multi_channel = True, bias = False)

and even more to:

        self.Pconv1 = PartialConv2d(3, 64, 7, 2, 3, multi_channel = True, bias = False)
        self.bn1 = nn.BatchNorm2d(64)
        self.Pconv2 = PartialConv2d(64, 64, 11, 1, 5, multi_channel = True, bias = False)
        self.bn20 = nn.BatchNorm2d(64)
        self.Pconv21 = PartialConv2d(64, 64, 11, 1, 5, multi_channel = True, bias = False)
        self.Pconv22 = PartialConv2d(64, 64, 11, 1, 5, multi_channel = True, bias = False)
        self.Pconv23 = PartialConv2d(64, 64, 3, 1, 1, multi_channel = True, bias = False)
        self.Pconv24 = PartialConv2d(64, 64, 3, 1, 1, multi_channel = True, bias = False)

However, I have not made this change because I fear it has other effects. I have not figured out what those are though.

**Comment by qwerdbeta** [comment]
Actually the larger kernal size (9 or 11) might get larger features which i want, but I'm not sure what the padding increases do or if they are necesary

**Comment by qwerdbeta** [comment]
What if we used larger kernal sizes and padding for the 2 new pconv layers? like this:

        self.Pconv1 = PartialConv2d(3, 64, 7, 2, 3, multi_channel = True, bias = False)
        self.bn1 = nn.BatchNorm2d(64)
        self.Pconv2 = PartialConv2d(64, 64, 7, 1, 3, multi_channel = True, bias = False)
        self.bn20 = nn.BatchNorm2d(64)
        self.Pconv21 = PartialConv2d(64, 64, 7, 1, 3, multi_channel = True, bias = False)
        self.Pconv22 = PartialConv2d(64, 64, 7, 1, 3, multi_channel = True, bias = False)
        self.Pconv23 = PartialConv2d(64, 64, 9, 1, 5, multi_channel = True, bias = False)
        self.Pconv24 = PartialConv2d(64, 64, 9, 1, 5, multi_channel = True, bias = False)

**Comment by qwerdbeta** [comment]
@Ad-Hua my other question is this: when you added Pconv23 and Pconv24, did you also add them to the forward function? Pconv21 and Pconv22 are added in the lines:

        for i in range(12):
            x2, m2 = self.Pconv21(x2, m2)
            x2, m2 = self.Pconv22(x2, m2)

**Comment by SSTTfit** [comment]
@qwerdbeta  Have you figured out the problem?About your two questions.

**Comment by mychina75** [comment]
I'v got this result. similiar with yours?
![96421393-4d9c9a80-1229-11eb-9e54-9486ed8f2a59_out](https://user-images.githubusercontent.com/798230/162555903-f66963ff-7cb3-43a4-8262-20f55b24d86b.png)
